server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -741,7 +741,12 @@ private void runReadyClone(
                         new ShardSnapshotUpdate(
                             target,
                             repoShardId,
-                            new ShardSnapshotStatus(localNodeId, ShardState.FAILED, "failed to clone shard snapshot", null)
+                            new ShardSnapshotStatus(
+                                localNodeId,
+                                ShardState.FAILED,
+                                "failed to clone shard snapshot",
+                                shardStatusBefore.generation()
```

Comment:
Can we shift some of the `@Nullable` annotations and associated assertions around in `ShardSnapshotStatus`? With this change I think we might still receive a null generation over the wire from an older version, but we shouldn't be creating them afresh any more?

Reply:
In fact we ought to change the wire format so it's no longer an `OptionalString`. I'm ok with not doing that here, it'll make the backport that much harder, a follow up is fine.

Reply:
We would in fact still create them for pre-7.6 state machine operation (still a thing if there's an old snapshot in your repo) and we don't have access to the snapshot version in these constructors. In these `null` means (figure out the numeric generation yourself which would happen to a queued operation if e.g. the first operation for a shard in the CS fails).

Let me see what I can do about this though :)

Reply:
Bad news here I'm afraid. I had to remove the assertion since it didn't hold up but it's also really hard to assert this elsewhere due to the BwC situation.
If it's ok with you I think I'd rather look for a cleaner way of asserting this stuff once #75501 has landed (or actually as part of incorporating this into that change) and just assert that we're not doing any illegal changes to `SnapshotsInProgress` like this any longer (much easier if we don't have to hack around translating `ShardId` and `RepoShardId` all over the place)?

Reply:
Sure, thanks for looking. Blasted BwC, always spoiling the fun.